### ISSUE TITLE
kube-dns restarts frequently - Exit Code 2

### ISSUE THREAD

**Issue by keir-rex** [kubernetes/kubernetes#48462]
BUG REPORT

/kind bug

**What happened**:
Kube DNS restarts very frequently with error message: `fatal error: concurrent map writes`

**What you expected to happen**:
Kube DNS should be more robust to failures and restart less frequently.

**How to reproduce it (as minimally and precisely as possible)**:
- Run `kube-dns-v15`
- We run a cluster of 3 Kube DNS nodes though this is unlikely required

**Anything else we need to know?**:
To my naive eye this could be related to issue #44711. If this is already fixed or negated by an update could you please recommend a compatible fix for Kubernetes version `1.5.2`? 

**Environment**:
- Kubernetes version (use `kubectl version`):
`Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}`
`Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+coreos.0", GitCommit:"3ed7d0f453a5517245d32a9c57c39b946e578821", GitTreeState:"clean", BuildDate:"2017-01-13T00:23:19Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}`
- Cloud provider or hardware configuration**: AWS 
- OS (e.g. from /etc/os-release): 
```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1298.5.0
VERSION_ID=1298.5.0
BUILD_ID=2017-02-28-0013
PRETTY_NAME="Container Linux by CoreOS 1298.5.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```
- Kernel (e.g. `uname -a`):
`Linux ip-xx-xx-xxx-xxx.ec2.internal 4.9.9-coreos-r1 #1 SMP Tue Feb 28 00:06:10 UTC 2017 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux`
- Install tools:
Custom installation using 'CoreOS: Kubernetes the hard way'
- Others:
Describe pods output and KubeDNS logs available [in this gist](https://gist.github.com/keir-rex/52031947af2ccab6b93d34ff90ed16cc)

**Comment by cmluciano** [comment]
I believe that https://github.com/kubernetes/dns/pull/79 was merged May 3rd, so it probably was not included in a release until 1.6/1.7. @MrHohn can you confirm?

**Comment by MrHohn** [comment]
That recent fix was cherrypicked into 1.6 branch (#47328) and should be released with v1.6.5+ (https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#changelog-since-v164).

@keir-rex Though you mentioned `Run kube-dns-v15` above, that is a really old version and the codes have changed dramatically since then.

ref: Below is the kubedns addon released with v1.5.2:
https://github.com/kubernetes/kubernetes/blob/v1.5.2/cluster/addons/dns/skydns-rc.yaml.base

**Comment by keir-rex** [comment]
Great, thanks for the update and pointers. I'll look at updating to the correct version for `1.5.2` and we're considering an upgrade to 1.6 in the near future.

Is there any chance of the fix for this race condition being back-ported to `1.5.x`? If not is it a bad idea to upgrade to the KubeDNS version released for `1.6.5`?

Finally, am I correct in my above assumption that this relates to #44711 - the concurrent write in my circumstance happens at `dns.go:453` whilst #44711 logs this happening at `dns.go:264`. Presumably, without diffing the versions, this may be due to refactoring?

Thanks again

**Comment by MrHohn** [comment]
@keir-rex Yes that should be due to refactoring. Because we moved kube-dns codes out of kubernetes/kubernetes repo in v1.6, and the fix is in kubernetes/dns repo, it would be tricky to backport fixes and images to v1.5.

I think it is safe to use kube-dns released for v1.6.5+ on v1.5 cluster. You will need to comment out the optional configMap part (it's a 1.6 feature) in template (https://github.com/kubernetes/kubernetes/blob/v1.6.6/cluster/addons/dns/kubedns-controller.yaml.base#L51-L55).

**Comment by cmluciano** [comment]
/close

please reopen if this persists on 1.6

**Comment by keir-rex** [comment]
Apologies for digging up this old issue. I'll open a new issue tomorrow if I don't get a reply.

After upgrading our environments qe, stage, production we began seeing nodes erroneously and seemingly at random begin to fail wrt DNS/Networking. This wasn't an issue we had encountered in the past. The networking on the host/node continues to work but all containers fail to resolve DNS.

We're running:
- `gcr.io/google_containers/kubedns-amd64:1.9` pulled from the k8s `1.6` release 
- Kubernetes `v1.5.2` cluster 
- Flannel `v0.6.2`
- Calico `v0.23.0`
- All running on CoreOS `1298.5.0`
 
My suspicion is that, from what I can tell, we're running CNI `0.3.1` bundled with Calico and k8s `1.6` requires CNI `0.5.1` as mentioned [here](https://github.com/coreos/flannel/blob/master/Documentation/Kubernetes.md).

We plan to upgrade ASAP but obviously want to know the underlying cause of this before we proceed.

The output of `kubectl describe pods` on our qe cluster is [here](https://gist.github.com/keir-rex/30981cd32ad46990aa76f662571d3a10)
